### PR TITLE
[NVIDIA XLA:GPU] Introducing training support for cudnn fused mha (stream executor, 1st out of 3)

### PR DESCRIPTION
**Pull request by Tixxx**
This is the commit to include only stream executor changes to support fmha training. In total, it's broken into 3 PRs.
This is the first of 3. The other PRs' links:
Rewriter changes are in this [pr](https://github.com/openxla/xla/pull/3726)
lhlo lowering logic is in this [pr](https://github.com/openxla/xla/pull/3886)

### CODE REVIEW

**Review by cheshire — APPROVED** [review]

**Review by cheshire — APPROVED** [review]

File: xla/mlir_hlo/lhlo_gpu/IR/lhlo_gpu_ops_enums.td [hunk 1]
```diff
@@ -137,6 +137,11 @@ def FusedMhaDagSoftmax : I32EnumAttrCase<"Softmax", 6>;
 def FusedMhaDagScaleBiasSoftmaxDropout : I32EnumAttrCase<"ScaleBiasSoftmaxDropout", 7>;
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/mlir_hlo/lhlo_gpu/IR/lhlo_gpu_ops_enums.td]: The latest commit removed this.

File: xla/service/gpu/backend_configs.proto [hunk 2]
```diff
@@ -120,18 +120,25 @@ message CudnnfMHABackendConfig {
...
   xla.DotDimensionNumbers bmm1_dot_dimension_numbers = 11;
   xla.DotDimensionNumbers bmm2_dot_dimension_numbers = 12;
 
   xla.ShapeProto intermediate_tensor_shape = 14;
 
+  // Configs for mha bmms in the backward graph
+  xla.DotDimensionNumbers bmm1_grad_gemm1_dot_dimension_numbers = 16;
+  xla.DotDimensionNumbers bmm1_grad_gemm2_dot_dimension_numbers = 17;
+  xla.DotDimensionNumbers bmm2_grad_gemm1_dot_dimension_numbers = 18;
+  xla.DotDimensionNumbers bmm2_grad_gemm2_dot_dimension_numbers = 19;
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/gpu/backend_configs.proto]: same comment as above

File: xla/stream_executor/dnn.cc [hunk 3]
```diff
@@ -250,6 +251,45 @@ DnnSupport::FusedMHAScaleBiasSoftmaxRunnerFromDesc(
...
+    const MatmulTensorDescriptor& bmm2_grad_gemm1_lhs_descriptor,
+    const MatmulTensorDescriptor& bmm2_grad_gemm2_rhs_descriptor,
+    const MatmulTensorDescriptor& d_output_descriptor,
+    const TensorDescriptor& d_bmm1_lhs_descriptor,
+    const TensorDescriptor& d_bmm1_rhs_descriptor,
+    const TensorDescriptor& d_bmm2_rhs_descriptor,
+    const TensorDescriptor& d_S_descriptor,
+    std::optional<dnn::TensorDescriptor> d_bias_descriptor, double scale,
+    std::optional<double> dropout_rate, std::optional<int64_t> seed) {
+  return tsl::errors::Unimplemented(
```
Comment (ddunl) [inline xla/stream_executor/dnn.cc:269]: For all the `tsl::errors::*` could you use the `absl` equivalent? like `absl::UnimplementedError` in this case

Comment (Tixxx) [reply xla/stream_executor/dnn.cc:269]: I pushed a new commit to address this, but somehow this pr doesnt reflect that commit..

Comment (ddunl) [reply xla/stream_executor/dnn.cc:269]: I noticed this myself, I got an email when you pushed the commit but couldn't see the changes. Not sure what's going on!

Comment (Tixxx) [reply xla/stream_executor/dnn.cc:269]: seems like it's still not showing up. I will do a rebase and see if that helps

Comment (Tixxx) [reply xla/stream_executor/dnn.cc]: there we go

Comment (ddunl) [reply xla/stream_executor/dnn.cc]: Looks like that worked, nice!

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 4]
```diff
@@ -4686,10 +4825,775 @@ GetCudnnFusedMHAOperationGraph(
           << "\nOpGraph: " << op_graph.describe();
```
Comment (ddunl) [inline xla/stream_executor/cuda/cuda_dnn.cc:4825]: Many of these seem to be use after moves now

Comment (Tixxx) [reply xla/stream_executor/cuda/cuda_dnn.cc:4831]: I have split up this vlog and move some of them befoe corresponding std::move

**Comment by ddunl** [comment]
This is good, I have to reapply my patch to get it submitted. Sorry this is taking me a bit, some fires to put out this morning

**Comment by Tixxx** [comment]
> This is good, I have to reapply my patch to get it submitted. Sorry this is taking me a bit, some fires to put out this morning

Thanks a lot!

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 5]
```diff
@@ -4675,21 +4824,782 @@ GetCudnnFusedMHAOperationGraph(
...
+  auto bmm1_grad_gemm1_op = cudnn_frontend::OperationBuilder(
+                                CUDNN_BACKEND_OPERATION_MATMUL_DESCRIPTOR)
+                                .setaMatDesc(tensor_ds_mask_reshape)
+                                .setbMatDesc(tensor_q)
+                                .setcMatDesc(tensor_dk)
+                                .setmatmulDesc(bmm1_grad_gemm1_desc)
+                                .build();
+  RETURN_MSG_IF_CUDNN_ERROR(bmm1_grad_gemm1_op);
+  intermdiate_ops.push_back(std::move(bmm1_grad_gemm1_op));
+
```
Comment (ddunl) [inline xla/stream_executor/cuda/cuda_dnn.cc:5543]: Possible to do `ops.reserve` here?

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 6]
```diff
@@ -4675,21 +4824,782 @@ GetCudnnFusedMHAOperationGraph(
...
+static tsl::StatusOr<cudnn_frontend::ExecutionPlan> GetExecPlanFromHeuristics(
+    cudnn_frontend::OperationGraph&& opGraph, const CudnnHandle& cudnn) {
+#if (CUDNN_VERSION >= 8800)
+  cudnn_frontend::EngineConfigList engine_configs;
+  cudnn_frontend::get_heuristics_list<1>({"heuristics_instant"}, opGraph,
+                                         allowAllConfig, engine_configs, true);
+  if (VLOG_IS_ON(4)) {
+    VLOG(4) << "Heuristic has " << engine_configs.size() << " configurations ";
+  }
+  if (engine_configs.size() == 0) {
```
Comment (ddunl) [inline xla/stream_executor/cuda/cuda_dnn.cc:5588]: Nit: I think could use `.empty()` instead (some other occurrences of this in the PR)

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 7]
```diff
@@ -4675,21 +4824,782 @@ GetCudnnFusedMHAOperationGraph(
...
+          << "\nBMM2_grad_gemm1: " << bmm2_grad_gemm1_desc.describe()
+          << "\nBMM2_grad_gemm1_op: " << bmm2_grad_gemm1_op.describe()
+          << "\nBMM2_grad_gemm2: " << bmm2_grad_gemm2_desc.describe()
+          << "\nBMM2_grad_gemm2_op: " << bmm2_grad_gemm2_op.describe()
+          << "\nBMM1_grad_gemm1: " << bmm1_grad_gemm1_desc.describe()
+          << "\nBMM1_grad_gemm1_op: " << bmm1_grad_gemm1_op.describe()
+          << "\nBMM1_grad_gemm2: " << bmm1_grad_gemm2_desc.describe()
+          << "\nBMM1_grad_gemm2_op: " << bmm1_grad_gemm2_op.describe()
+          << "\nOpGraph: " << op_graph.describe();
+  return std::unique_ptr<cudnn_frontend::OperationGraph>(
```
Comment (ddunl) [inline xla/stream_executor/cuda/cuda_dnn.cc:5571]: I think this could be std::make_unique

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 8]
```diff
@@ -4675,21 +4824,782 @@ GetCudnnFusedMHAOperationGraph(
...
+          << "\nTensor_dk: " << tensor_dk.describe()
+          << "\nTensor_dv: " << tensor_dv.describe()
+          << "\nBMM2_grad_gemm1: " << bmm2_grad_gemm1_desc.describe()
+          << "\nBMM2_grad_gemm1_op: " << bmm2_grad_gemm1_op.describe()
+          << "\nBMM2_grad_gemm2: " << bmm2_grad_gemm2_desc.describe()
+          << "\nBMM2_grad_gemm2_op: " << bmm2_grad_gemm2_op.describe()
+          << "\nBMM1_grad_gemm1: " << bmm1_grad_gemm1_desc.describe()
+          << "\nBMM1_grad_gemm1_op: " << bmm1_grad_gemm1_op.describe()
+          << "\nBMM1_grad_gemm2: " << bmm1_grad_gemm2_desc.describe()
+          << "\nBMM1_grad_gemm2_op: " << bmm1_grad_gemm2_op.describe()
```
Comment (ddunl) [inline xla/stream_executor/cuda/cuda_dnn.cc:5569]: Use after move here

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 9]
```diff
@@ -4675,21 +4824,782 @@ GetCudnnFusedMHAOperationGraph(
...
+          << "\nTensor_do: " << tensor_do.describe()
+          << "\nTensor_dq: " << tensor_dq.describe()
+          << "\nTensor_dk: " << tensor_dk.describe()
+          << "\nTensor_dv: " << tensor_dv.describe()
+          << "\nBMM2_grad_gemm1: " << bmm2_grad_gemm1_desc.describe()
+          << "\nBMM2_grad_gemm1_op: " << bmm2_grad_gemm1_op.describe()
+          << "\nBMM2_grad_gemm2: " << bmm2_grad_gemm2_desc.describe()
+          << "\nBMM2_grad_gemm2_op: " << bmm2_grad_gemm2_op.describe()
+          << "\nBMM1_grad_gemm1: " << bmm1_grad_gemm1_desc.describe()
+          << "\nBMM1_grad_gemm1_op: " << bmm1_grad_gemm1_op.describe()
```
Comment (ddunl) [inline xla/stream_executor/cuda/cuda_dnn.cc:5567]: Use after move here

File: xla/stream_executor/cuda/cuda_dnn.h [hunk 10]
```diff
@@ -324,6 +324,38 @@ class CudnnSupport : public dnn::DnnSupport {
...
+      const dnn::MatmulTensorDescriptor& bmm1_grad_gemm2_rhs_descriptor,
+      const dnn::MatmulTensorDescriptor& bmm2_grad_gemm1_lhs_descriptor,
+      const dnn::MatmulTensorDescriptor& bmm2_grad_gemm2_rhs_descriptor,
+      const dnn::MatmulTensorDescriptor& d_output_descriptor,
+      const dnn::TensorDescriptor& d_bmm1_lhs_descriptor,
+      const dnn::TensorDescriptor& d_bmm1_rhs_descriptor,
+      const dnn::TensorDescriptor& d_bmm2_rhs_descriptor,
+      const dnn::TensorDescriptor& d_S_descriptor,
+      std::optional<dnn::TensorDescriptor> d_bias_descriptor, double scale,
+      std::optional<double> dropout_rate, std::optional<int64_t> seed);
```
Comment (ddunl) [inline xla/stream_executor/cuda/cuda_dnn.h:341]: Should have override I think?

File: xla/stream_executor/cuda/cuda_dnn.h [hunk 11]
```diff
@@ -324,6 +324,38 @@ class CudnnSupport : public dnn::DnnSupport {
...
+      const dnn::MatmulTensorDescriptor& bmm2_grad_gemm1_lhs_descriptor,
+      const dnn::MatmulTensorDescriptor& bmm2_grad_gemm2_rhs_descriptor,
+      const dnn::MatmulTensorDescriptor& d_output_descriptor,
+      const dnn::TensorDescriptor& d_bmm1_lhs_descriptor,
+      const dnn::TensorDescriptor& d_bmm1_rhs_descriptor,
+      const dnn::TensorDescriptor& d_bmm2_rhs_descriptor,
+      const dnn::TensorDescriptor& d_S_descriptor,
+      const dnn::TensorDescriptor& mask_descriptor,
+      std::optional<dnn::TensorDescriptor> d_bias_descriptor, double scale,
+      std::optional<double> dropout_rate, std::optional<int64_t> seed);
```
Comment (ddunl) [inline xla/stream_executor/cuda/cuda_dnn.h:358]: Should have override I think as well

File: xla/stream_executor/dnn.cc [hunk 12]
```diff
@@ -246,10 +247,49 @@ DnnSupport::FusedMHAScaleBiasSoftmaxRunnerFromDesc(
...
+    dnn::FusedMHAKind kind,
+    const MatmulTensorDescriptor& bmm1_grad_gemm1_rhs_descriptor,
+    const MatmulTensorDescriptor& bmm1_grad_gemm2_rhs_descriptor,
+    const MatmulTensorDescriptor& bmm2_grad_gemm1_lhs_descriptor,
+    const MatmulTensorDescriptor& bmm2_grad_gemm2_rhs_descriptor,
+    const MatmulTensorDescriptor& d_output_descriptor,
+    const TensorDescriptor& d_bmm1_lhs_descriptor,
+    const TensorDescriptor& d_bmm1_rhs_descriptor,
+    const TensorDescriptor& d_bmm2_rhs_descriptor,
+    const TensorDescriptor& d_S_descriptor,
```
Comment (ddunl) [inline xla/stream_executor/dnn.cc:266]: Optional: the style guide complains about this, it wants this to be `d_s_descriptor`, but it occurs to me the S might be getting at some standard notation I'm not aware of. Whatever you think is best here works for me

**Review by ddunl — COMMENTED** [review]
Sorry for the delay, after this it should be good for sure. I'm not sure how I missed these last time, my bad on that

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 7 above)
Comment (Tixxx) [reply xla/stream_executor/cuda/cuda_dnn.cc]: changed

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 5 above)
Comment (Tixxx) [reply xla/stream_executor/cuda/cuda_dnn.cc:5556]: added reserve, also changed push_back to emplace_back

File: xla/stream_executor/dnn.cc (hunk 12 above)
Comment (Tixxx) [reply xla/stream_executor/dnn.cc]: changed all capital to lower case for this variable

**Pull request closed without merging**